### DISCUSSION TITLE
Plugin Help

### DISCUSSION THREAD

**Discussion by rvalotta** [grblHAL/plugins#1] [Q&A]
Any chance someone could help me with development of a plugin that simply does the same thing a $32=1. Would like to be able to change in and out of laser mode without having to update flash.

**Comment by terjeio** [comment]
A plugin adding a M-code for switching spindle is easy to write, how to communicate which one to select to spindle driver code is what is a bit challenging. One way to do it is to have all spindle drivers implement a M-code but that is not very flexible and will lead to code duplication.

My idea is to add an event for spindle change to the [core handlers](https://github.com/grblHAL/core/blob/master/core_handlers.h) that spindle drivers will have to subscribe to:

`typedef bool (*on_spindle_select_ptr)(uint_fast8_t spindle_id);`

A plugin implementing a M-code can then request a spindle change like this:

```
if(!(grbl.on_spindle_select && grbl.on_spindle_select(spindle_id))
     raise error?
```
where `spindle_id` is extracted from a M-code parameter.

`spindle_id 0` is for selecting PWM output and will be the default when laser mode is enabled, `1` will be the default if not.

I need a bit of time to add this to the core and modify/test the Huanyang plugin...

**Reply by terjeio** [reply]
Can you try to rat out which M-code to use (if any) from how other systems implement spindle selection? I have done a few searches without any success...

**Reply by rvalotta** [reply]
Same here. I will start going through linuxcnc and marlin's gcode listing and see what i can find....  Thank you for this. Here I thought all I would have to do is set lasermode=1 in code :)

**Reply by rvalotta** [reply]
The more I think about this, is it not just a matter of turning on laser mode. We wouldn't be using M3/M4/M5 to turn on and off the laser that would happen automagically with the G0 / G1 right?   If laser_mode use spindle_pwm (Maybe call is laser_pwm)  otherwise use vfd?   I'm sure that is wrong but I have yet to wrap my head on how this all ties together. I am working at that though.

**Reply by terjeio** [reply]
> Here I thought all I would have to do is set lasermode=1 in code

Yes, but you cannot set the settings value without storing it to flash, and the spindle driver has to be told about the change so it can take appropriate action. The spindle driver will switch off the current spindle and change the speed to 0. After switching the selected spindle has to be turned on and speed set anew. And a new offset has to be selected and moved to as well as both spindles do not have the same XY position?

> Same here. I will start going through linuxcnc and marlin's gcode listing and see what i can find.... 

I could not find anything in the LinuxCNC or Marlin documentation. I know that LinuxCNC supports multiple spindles and I believe I have seen somwhere how spindle selection is done, was it by a specially formatted message? AFAIK Marlin supports selection of extruder, but again I could not find how it is done.

There are CNC machines available that supports multiple spindles, but again I did not find out how selection is performed? Perhaps by a relay and a physical switch?

**Reply by rvalotta** [reply]
Found this last night showing how multi spindle is handled by LinuxCNC

https://forum.linuxcnc.org/24-hal-components/35370-secondary-spindle-hal-integration

**Reply by terjeio** [reply]
> Found this last night showing how multi spindle is handled by LinuxCNC

Thanks, this implementation is odd and not really suitable as it is for enabling several spindles at the same time, and uses `$` as a parameter letter. The `$` may confuse senders, at least ioSender is likely to report an error on load (I have not checked).
Some Haas lathes uses `G14` for selecting spindle in dual spindle lathes, but is not well suited either as it mirrors the job...

So I have decided to hijack `M104` for the time beeing, this is user definable M-code for grblHAL.
`M104P0` for selecting the PWM spindle, `M104P1` for the VFD spindle - available when `DUAL_SPINDLE` option is compiled in.
I will commit an update with this soon - for testing.

**Reply by rvalotta** [reply]
Saw you pushed this going to be a few days before i can wire in the laser I will update you as soon as I've got it all connected

**Reply by rvalotta** [reply]
I did do some benchtop testing and it appears to be working

**Reply by rvalotta** [reply]
Finally got to spend some time and test this... Its working very well thankyou. My only request is there a way to have it do what ever $32=1 would do (IE set it to laser mode) without having to the $32=1 (Right now the beam is staying on with G0 moves)

**Reply by terjeio** [reply]
>  My only request is there a way to have it do what ever $32=1 would do (IE set it to laser mode) without having to the $32=1

This will require a core change, and the best option would be to set $31=1 and switch to normal mode when the VFD is enabled.

**Reply by terjeio** [reply]
In the next commit I will add [a new variable](http://svn.io-engineering.com/grblHAL/html/structsystem.html), `sys.mode`, to control the mode without changing the $32 setting. This can be modified on the fly in a new plugin, or perhaps in VFD plugins, as required.
